### ISSUE TITLE
Payment untaken in Stripe

### ISSUE THREAD

**Issue by Cecilia-Hn** [openfoodfoundation/openfoodnetwork#8218]
## Description

We've got one case of a payment untaken in Stripe after an order thrugh OFN : 

A client selected some product in a store 

![Screenshot_20210921-112354_Chrome(1)](https://user-images.githubusercontent.com/67476305/134196654-516dbd83-6258-4018-bf38-fa3fde378538.jpg)

Then, the client tried to paid his order with his bank card : 
- he was redirected on his bank app where the payment has been validated 
- then he was redirected on OFN where this error message appeared 

![Screenshot_20210921-112716_Chrome(1)](https://user-images.githubusercontent.com/67476305/134197133-7a34d3a0-063b-493f-b0f1-772cf291c2ac.jpg)

## Expected Behavior
Payment should be captured by stripe and then by OFN to allow client to take orders. 

## Actual Behaviour

The client made two attempts.

First at 9.20

<img width="612" alt="Capture d’écran 2021-09-21 à 17 14 43" src="https://user-images.githubusercontent.com/67476305/134198144-57bd71f7-b4ef-48dd-be3e-c6f502da9f96.png">

Second at 9:26

<img width="655" alt="Capture d’écran 2021-09-21 à 17 15 40" src="https://user-images.githubusercontent.com/67476305/134198326-9915d587-e364-4516-9b16-58843cb77d9f.png">

## Workaround
Not really work around here

**Comment by filipefurtad0** [comment]
Hey @Cecilia-Hn ,

Here's what I could gather:

There are 14 (!) payment attempts on this order (order number = R353458575 // order_id = 2336323):
```
openfoodnetwork=> select count(*) from spree_payments where order_id = 2336323;
 count 
-------
    14
(1 row)
```

 Only the first one was successful:

```
   id   | amount | order_id |         created_at         |         updated_at         | source_id |    source_type    | payment_method_id |         state          |        response_code        | avs_response | identifier | cvv_response_code |                                                                                                                                                           cvv_response_message                                                                                                                                                            
--------+--------+----------+----------------------------+----------------------------+-----------+-------------------+-------------------+------------------------+-----------------------------+--------------+------------+-------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
186292 |  58.00 |  2336323 | 2021-09-20 13:08:07.368825 | 2021-09-20 13:08:11.735344 |     27303 | Spree::CreditCard |               238 | completed              | pi_3JbmE8EtcoBMEZn802GqGF8O |              | CJCPNYBU   |  
```

All 13 attempts after this one either failed, pending, or require authentication. One of them is the one from 09:26:03 AM:
```
   id   | amount | order_id |         created_at         |         updated_at         | source_id |    source_type    | payment_method_id |         state          |        response_code        | avs_response | identifier | cvv_response_code |                                                                                                                                                           cvv_response_message                                                                                                                                                            
--------+--------+----------+----------------------------+----------------------------+-----------+-------------------+-------------------+------------------------+-----------------------------+--------------+------------+-------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 186563 | 106.00 |  2336323 | 2021-09-21 09:26:03.056545 | 2021-09-21 09:26:06.056213 |     27368 | Spree::CreditCard |               238 | requires_authorization | pi_3Jc5EmEtcoBMEZn80a4pDGaI |              | Y5WUNQFD   |                   | https://hooks.stripe.com/3d_secure_2/hosted?merchant=acct_1DyLlqEtcoBMEZn8&payment_intent=pi_3Jc5EmEtcoBMEZn80a4pDGaI&payment_intent_client_secret=pi_3Jc5EmEtcoBMEZn80a4pDGaI_secret_QCKOefaod70bGF9geH95gTlcK&publishable_key=pk_live_ZLcHKa6qfIroIekloY2l0i1D&source=src_1Jc5EnEtcoBMEZn8L4werGp3&stripe_account=acct_1DyLlqEtcoBMEZn8
```

Which triggered this Bugsnag error:

https://app.bugsnag.com/open-food-france/coopcircuits/errors/60f11496903af700075b686d?filters[event.since]=all&event_id=6149a5700080da1e77030000

No sure what led to this scenario...

**Comment by Cecilia-Hn** [comment]
The untaken payment by Stripe concerned the all following hub : 

- ID OFN : 1265 
- ID OFN : 2236 

I think way more are concerned.

**Comment by filipefurtad0** [comment]
An error appeared today which could be related to this issue (checkout + Stripe), although the error message seems different:
https://app.bugsnag.com/open-food-france/coopcircuits/errors/60d1ff0db5a60f0008d2b0f1?event_id=614b4eb6008122268e380000&i=sk&m=fq

**Comment by Cecilia-Hn** [comment]
Hi 

I just had the concerned hub on the phone, they were also impacted about the result of back office settings in shipping methods until monday. Maybe this is linked ?

**Comment by filipefurtad0** [comment]
Thanks for mentioning that @Cecilia-Hn , I've checked that scenario and indeed they look related; [I've posted on #8219](https://github.com/openfoodfoundation/openfoodnetwork/issues/8219#issuecomment-925270832) , so the dev which picks this up is aware of possible steps to reproduce.

**Comment by mkllnk** [comment]
A good hint is the error message:

> The payment could not be captured because the requested capture amount is greater than the authorized amount.

And from the history above it looks like the first payment had a different amount to the following payments.

One theory is that we somehow authorised for $58 but then tried to charge for $106. But before all that, we need to find out why the first payment got marked as completed in our system while Stripe didn't take it.

**Comment by mkllnk** [comment]
I had a closer look at Bugsnag and [this error](https://app.bugsnag.com/open-food-france/coopcircuits/errors/60f11496903af700075b686d) happened only in France and only to that one user.

**Comment by mkllnk** [comment]
There are 14 payments. The first one was successful in OFN but the other 13 are either `pending`, `requires_authorization` or `failed`.

We got 12 Bugsnag error for that user around that time. The first five have timestamps that directly correlate to the first five failed payments. After that, it looks like some payments have been updated after the error was logged. Most error messages are:

> Le paiement n'a pas pu être traité, veuillez vérifier les informations saisies
> (The payment could not be processed, please check the information entered)

That sounds like a user error, for example a typo in the card number. Only the last two payments failed with:

> The payment could not be captured because the requested capture amount is greater than the authorized amount. If you are trying to collect a tip or over-capture, your account might not be configured for those features. Contact us via https://support.stripe.com/contact/

The order actually changed after the first completed payment. Two of the three line items are older than the first payment and that may explain why payment amounts vary. They are either EUR 58 (to start with) or EUR 106 (towards the end). Interestingly, some payments were created while the order still had a value of 58 and were then updated to 106 later. That could explain the error message above.

**Comment by mkllnk** [comment]
Theories to confirm:

* The first payment was successful but another reason prevented the checkout, for example stock issues.
* I haven't checked Stripe yet if that first payment is actually displayed as failed.
* Was the first successful payment then voided because the checkout failed but OFN didn't update the local database?
* The user changed the cart after the first payment. Confirmed.
* The user could have changed the cart during the authorisation step.
* Later payment attempts were for the full cart amount (106) and not just for the remaining EUR 46 which would "overpay" the order and that may cause an error within OFN.
* The authorisation step is triggered from Javascript in the checkout form and the total may be outdated if the cart was modified in another tab. The final charge may then have been for a different amount, triggered from the server side.

**Comment by RachL** [comment]
Thanks @mkllnk that's super helpful. Shall we consider that this is an edge case and stop here?

I guess that https://github.com/openfoodfoundation/openfoodnetwork/issues/7961 will help us personalize those messages and include a line around not trying to checkout on multiple tab?

**Comment by andrewpbrett** [comment]
Given that this happened in the shop that had their shipping methods set to backoffice only (if I'm understanding @Cecilia-Hn 's [comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/8218#issuecomment-925116635) correctly), it seems like we should treat this issue as a side effect of that one and close this issue. Thanks to @mkllnk's investigations I think it's pretty clear this is an edge case so I think that the S3 label for https://github.com/openfoodfoundation/openfoodnetwork/issues/8219 is correct. 

Feel free to re-open, anyone, if I missed something or you disagree.

**Comment by Cecilia-Hn** [comment]
Hi, I jus had these info from the customer, maybe it can helps @mkllnk @filipefurtad0 @andrewpbrett  : 

- I did a first order for 58 euros : few error messages appeared while I was trying to order (at least 5 times) so, then I gave up.
- The day after, I completed my first 58 euros order for 106 euros, then I did two or maximum 3 tried and then, I gave up.
- The day after the day after I realised that my bank account has been withdrawm of 58 euros

I ordered from my phone with only one OFN taf oppened. 

My internet connexion was ADSL so not fabulous bu I ordered from other shops without problems.
